openstudiocore/src/model/BuildingUnit.cpp

```diff
@@ -0,0 +1,430 @@
+      } else {
+        value = boost::none;
+      }
+    } else {
+      value = boost::none;
+    }
+    return value;
+  }
+
+  boost::optional<std::string> BuildingUnit_Impl::getFeatureAsString(const std::string& name) const
+  {
+    return getFeatureStringAndCheckForType(name, "alpha");
+  }
+
+  boost::optional<double> BuildingUnit_Impl::getFeatureAsDouble(const std::string& name) const
+  {
+    boost::optional<std::string> strValue(getFeatureStringAndCheckForType(name, "real"));
+    boost::optional<double> value;
+    if (strValue) {
+      try {
+        value = boost::lexical_cast<double>(*strValue);
+      } catch (boost::bad_lexical_cast) {
+        LOG(Error, "Value: " + *strValue + ", not castable to type double.")
+        value = boost::none;
+      }
+    } else {
+      value = boost::none;
+    }
+    return value;
+  }
+
+  boost::optional<int> BuildingUnit_Impl::getFeatureAsInteger(const std::string& name) const
+  {
+    boost::optional<std::string> strValue(getFeatureStringAndCheckForType(name, "integer"));
+    boost::optional<int> value;
+    if (strValue) {
+      try {
+        value = boost::lexical_cast<int>(*strValue);
+      } catch (boost::bad_lexical_cast) {
+        LOG(Error, "Value: " + *strValue + ", not castable to type integer.")
+        value = boost::none;
+      }
+    } else {
+      value = boost::none;
+    }
+    return value;
+  }
+
+  boost::optional<bool> BuildingUnit_Impl::getFeatureAsBoolean(const std::string& name) const
+  {
+    boost::optional<std::string> strValue(getFeatureStringAndCheckForType(name, "boolean"));
+    boost::optional<bool> value;
+    if (strValue) {
+      if (*strValue == "false") {
+        value = false;
+      } else if (*strValue == "true") {
+        value = true;
+      } else {
+        LOG(Error, "Value: " + *strValue + ", not castable to type boolean.")
+        value = boost::none;
+      }
+    } else {
+      value = boost::none;
+    }
+    return value;
+  }
+
+  std::vector<std::string> BuildingUnit_Impl::suggestedFeatures() const
+  {
+    std::set<std::string> availableFeatureNames;
+    availableFeatureNames.insert("NumberOfBedrooms");
+    availableFeatureNames.insert("NumberOfBathrooms");
+    for (const BuildingUnit& bldgUnit : this->model().getConcreteModelObjects<BuildingUnit>()) {
+      for (const std::string& featureName : bldgUnit.featureNames()) {
+        availableFeatureNames.insert(featureName);
+      }
+    }
+    std::vector<std::string> retvals;
+    retvals.assign(availableFeatureNames.begin(), availableFeatureNames.end());
+    return retvals;
+  }
+
+  bool BuildingUnit_Impl::setFeatureGroupDataTypeAndValue(const std::string& name, const std::string& dataType, const std::string& value)
+  {
+    boost::optional<ModelExtensibleGroup> group(getFeatureGroupByName(name));
+    if (group) {
+      const bool dataTypeOK = group->setString(OS_BuildingUnitExtensibleFields::BuildingUnitFeatureDataType, dataType);
```

Comment:
also, should be aware that signals will fire after each setString, any slots attached to that signal may fire after the first set and get an out of state object.  There is a badly named third argument to setString called checkValidity, if that is passed in as false the setter will not emit change signals.:

https://github.com/NREL/OpenStudio/blob/os_2_0_develop/openstudiocore/src/utilities/idf/WorkspaceObject_Impl.hpp#L195